### DISCUSSION TITLE
Vtm online map servers

### DISCUSSION THREAD

**Discussion by Psijic** [mapsforge/vtm#1059] [Q&A] [answered]
Tried to use Vtm with online public servers but didn't find any free. Is there a list or it's a crap idea? Looks like `UrlTileSource` need to be used, but found only `OSciMap4TileSource` and custom `MapnikVectorTileSource` classes for it.
Tried with this sample code  but looks like all is outdated and URLs dumped. Got only scale bar and gray background rendered.

```
@Composable
fun VtmMapView() {
    val mapView = rememberMapViewWithLifecycle()
    with (mapView.map()){
        // Scale bar
        val mapScaleBar: MapScaleBar = DefaultMapScaleBar(mapView.map())
        val mapScaleBarLayer = MapScaleBarLayer(mapView.map(), mapScaleBar)
        mapScaleBarLayer.renderer.setPosition(GLViewport.Position.BOTTOM_LEFT)
        mapScaleBarLayer.renderer.setOffset(5 * CanvasAdapter.getScale(), 0f)
        mapView.map().layers().add(mapScaleBarLayer)

        val tileSource = OSciMap4TileSource() // or MapnikVectorTileSource()

        // Vector layer
        val tileLayer: VectorTileLayer = mapView.map().setBaseMap(tileSource)

        // Building layer
        layers().add(BuildingLayer(mapView.map(), tileLayer))

        // Label layer
        layers().add(LabelLayer(mapView.map(), tileLayer))

        setMapPosition(52.517037, 13.38886, 12.0)
    }

    AndroidView(factory = { mapView }) {}
}

class MapnikVectorTileSource : UrlTileSource("https://d1s11ojcu7opje.cloudfront.net/dev/764e0b8d", "") {
    init {
        urlFormatter = TileUrlFormatter { tileSource, tile -> // url formatter for mapbox streets
            val hexTable = byteArrayOf(
                '0'.code.toByte(), '1'.code.toByte(), '2'.code.toByte(), '3'.code.toByte(),
                '4'.code.toByte(), '5'.code.toByte(), '6'.code.toByte(), '7'.code.toByte(),
                '8'.code.toByte(), '9'.code.toByte(), 'a'.code.toByte(), 'b'.code.toByte(),
                'c'.code.toByte(), 'd'.code.toByte(), 'e'.code.toByte(), 'f'
                    .code.toByte()
            )
            val sb = StringBuilder()
            sb.append('/')
            sb.append(hexTable[tile.tileX % 16].toInt())
            sb.append(hexTable[tile.tileY % 16].toInt())
            sb.append('/')
            sb.append(tile.zoomLevel.toInt())
            sb.append('/')
            sb.append(tile.tileX)
            sb.append('/')
            sb.append(tile.tileY)
            sb.toString()
        }
    }

    override fun getDataSource(): ITileDataSource {
        return UrlTileDataSource(this, TileDecoder(), httpEngine)
    }
}
```

**Comment by Psijic** [comment]
Chatbot says there are some. Is there any for free usage?

There are a few options for online vector tile servers that can be used with mapsforge and VTM:

OpenMapTiles
Vector tiles hosted by OpenMapTiles
Global coverage
https://openmaptiles.com/
Example URL: [https://free-{subDomain}.tilehosting.com/data/v3/{z}/{x}/{y}.pbf](https://free-%7Bsubdomain%7D.tilehosting.com/data/v3/%7Bz%7D/%7Bx%7D/%7By%7D.pbf)

OpenStreetMap Vector Tiles
Vector tiles generated from OpenStreetMap data
Global coverage
https://osm-vector-tiles.tileserver.ch/
Example URL: [https://osm-vector-tiles.tileserver.ch/data/v3/{z}/{x}/{y}.pbf](https://osm-vector-tiles.tileserver.ch/data/v3/%7Bz%7D/%7Bx%7D/%7By%7D.pbf)

MapTiler Cloud
Freemium vector tiles with Mapsforge maps
Global coverage
https://www.maptiler.com/cloud/
Example URL: [https://api.maptiler.com/tiles/v3/{z}/{x}/{y}.pbf?key={apiKey}](https://api.maptiler.com/tiles/v3/%7Bz%7D/%7Bx%7D/%7By%7D.pbf?key=%7BapiKey%7D)

Nextzen Vector Tiles
Global vector tiles with building footprints
Free low volume usage
https://www.nextzen.org/documentation/vector-tiles/
Example URL: [https://tile.nextzen.org/tilezen/vector/v1/all/{z}/{x}/{y}.mvt](https://tile.nextzen.org/tilezen/vector/v1/all/%7Bz%7D/%7Bx%7D/%7By%7D.mvt)

Mapbox Vector Tiles
Requires API key
High detail data with custom styling
https://www.mapbox.com/vector-tiles

**Comment by devemux86** [comment, marked as answer]
See also the samples, like OpenMapTilesMvtActivity, MapilionMvtActivity.

**Comment by Psijic** [comment]
Tried `val tileSource = OSciMap4TileSource("https://tile.nextzen.org/tilezen/vector/v1/256/all/{Z}/{X}/{Y}.mvt?api_key=my_key")` but it doesn't work. Are these tiles supported?

**Reply by devemux86** [reply]
Some servers may no longer work.

Also study the examples, there is NextzenMvtActivity with NextzenMvtTileSource.

**Reply by Psijic** [reply]
Surprisingly, it's not included in the main library.
Tried this code to implement but failed:
```
class NextzenTileSource(
    apiKey: String,
    urlString: String = "https://tile.nextzen.org/tilezen/vector/v1/all",
    tilePath: String = "{z}/{x}/{y}.mvt"
) : UrlTileSource(urlString, tilePath) {

    init {
        setApiKey(apiKey)
    }

    override fun getDataSource(): ITileDataSource {
        return UrlTileDataSource(this, TileDecoder(), httpEngine)
    }

    override fun getTileUrl(tile: Tile): String {
        return "${url}/${tile.zoomLevel}/${tile.tileX}/${tile.tileY}.mvt"
    }

}
```
Will try to convert that java class.

**Reply by devemux86** [reply]
It cannot because it requires extra dependencies.

**Reply by Psijic** [reply]
I copied this class to my project and only added 1 import: `import org.oscim.tiling.source.mapzen.TileDecoder;`

**Comment by Psijic** [comment]
Dat Java. Is it correct implementation? I see no errors in Logcat.
`val tileSource = NextzenMvtTileSource.builder().url("https://tile.nextzen.org/tilezen/vector/v1/256/all/{Z}/{X}/{Y}.mvt").apiKey(apiKey).build()`
Also tried to use only the key cause URL is already implemented.
`val tileSource = NextzenMvtTileSource.builder().apiKey(apiKey).build()`

Maybe the server is down:
_Nextzen is experimenting with a lower cost cached mode. Please contact hello@nextzen.org with questions._

**Comment by Psijic** [comment]
Somehow launched Android sample with working maps: Simple map works only with few Berlin areas and high scale. 
Nextzen miss some areas/scales partially, like tiles download by UDP with drops. Some scales doesn't show labels in Europe at all except UK London city.
Mapilion works quite smoothly especially disabling extra shadows. But it has a restricted key and free commercial keys are unavailable.
Cache is disabled.
